semgrep-core/parsing/Parse_rust_tree_sitter.ml

```diff
@@ -339,2122 +389,2262 @@ let map_macro_rule (env : env) ((v1, v2, v3) : CST.macro_rule) =
-      let v2 =
-        (match v2 with
-        | Some tok -> token env tok (* "mut" *)
-        | None -> todo env ())
-      in
-      let v3 = map_expression env v3 in
-      todo env (v1, v2, v3)
-  | `Try_exp (v1, v2) ->
-      let v1 = map_expression env v1 in
-      let v2 = token env v2 (* "?" *) in
-      todo env (v1, v2)
-  | `Bin_exp x -> map_binary_expression env x
-  | `Assign_exp (v1, v2, v3) ->
-      let v1 = map_expression env v1 in
-      let v2 = token env v2 (* "=" *) in
-      let v3 = map_expression env v3 in
-      todo env (v1, v2, v3)
-  | `Comp_assign_expr (v1, v2, v3) ->
-      let v1 = map_expression env v1 in
-      let v2 =
-        (match v2 with
-        | `PLUSEQ tok -> token env tok (* "+=" *)
-        | `DASHEQ tok -> token env tok (* "-=" *)
-        | `STAREQ tok -> token env tok (* "*=" *)
-        | `SLASHEQ tok -> token env tok (* "/=" *)
-        | `PERCEQ tok -> token env tok (* "%=" *)
-        | `AMPEQ tok -> token env tok (* "&=" *)
-        | `BAREQ tok -> token env tok (* "|=" *)
-        | `HATEQ tok -> token env tok (* "^=" *)
-        | `LTLTEQ tok -> token env tok (* "<<=" *)
-        | `GTGTEQ tok -> token env tok (* ">>=" *)
-        )
-      in
-      let v3 = map_expression env v3 in
-      todo env (v1, v2, v3)
-  | `Type_cast_exp (v1, v2, v3) ->
-      let v1 = map_expression env v1 in
-      let v2 = token env v2 (* "as" *) in
-      let v3 = map_type_ env v3 in
-      todo env (v1, v2, v3)
-  | `Range_exp x -> map_range_expression env x
-  | `Call_exp (v1, v2) ->
-      let v1 = map_expression env v1 in
-      let v2 = map_arguments env v2 in
-      todo env (v1, v2)
-  | `Ret_exp x -> map_return_expression env x
-  | `Lit x -> map_literal env x
-  | `Id tok ->
-      token env tok (* pattern (r#)?[a-zA-Zα-ωΑ-Ωµ_][a-zA-Zα-ωΑ-Ωµ\d_]* *)
-  | `Choice_u8 x -> map_anon_choice_u8_6dad923 env x
-  | `Choice_defa x -> map_reserved_identifier env x
-  | `Self tok -> token env tok (* "self" *)
-  | `Scoped_id x -> map_scoped_identifier env x
-  | `Gene_func (v1, v2, v3) ->
-      let v1 =
-        (match v1 with
+   | `Un_exp (v1, v2) ->
+       let expr = map_expression env v2 in
+       (match v1 with
+        | `DASH tok -> let tok = token env tok in (* "-" *)
+            G.Call (G.IdSpecial (G.Op G.Minus, tok), fb [G.Arg expr;])
+        | `STAR tok -> let tok = token env tok in (* "*" *)
+            G.DeRef (tok, expr)
+        | `BANG tok -> let tok = token env tok in (* "!" *)
+            G.Call (G.IdSpecial (G.Op G.Not, tok), fb [G.Arg expr;])
+       )
+   | `Ref_exp (v1, v2, v3) ->
+       let ref_ = token env v1 (* "&" *) in
+       let mutability = Option.map (fun tok ->
+         let tok = token env tok (* "mut" *) in
+         G.KeywordAttr (G.Mutable, tok)
+       ) v2 in
+       let expr = map_expression env v3 in
+       G.Ref (ref_, expr)
+   | `Try_exp (v1, v2) ->
+       let v1 = map_expression env v1 in
+       let v2 = token env v2 (* "?" *) in
+       todo env (v1, v2)
```

Comment:
could be a Call IdSpecial with Nullish. I think Typescript or Javascript have a similar operator.

Reply:
The semantics are different, it's more like a try-catch operator using Rust's `Result<T, E>` type. It replaces this

```rust
    let f = File::open("username.txt");
    let mut f = match f {
        Ok(file) => file,
        Err(e) => return Err(e),
    };
```

with this

``` rust
   let file = File::open("username.txt")?;
```

Reply:
I see. There's a related thing in Kotlin with ?.  https://kotlinlang.org/docs/reference/null-safety.html
We might introduce indeed a new operator in AST_generic.ml for that. Not sure about a good name.

Reply:
It's called the [safe navigation operator](https://en.wikipedia.org/wiki/Safe_navigation_operator).

Perhaps `SafeNavigation`?

Reply:
Actually it's already in the code. It's called `Elvis`, but Kotlin's `?` doesn't actually perform the same function as the elvis operator. I'll just use that for now.

Reply:
Yes, it does not have to be exactly the same semantic. Right now we don't do any deep analysis about those operators, so as long as you don't use the same operator for 2 different constructs in the same language, you're fine. For example it's ok to use TyPointer in Rust for &, because there is no * in Rust, but for C++ we need to introduce a new TyRef because we don't want *x and &x to match to the same generic AST construct.

Reply:
> there is no * in Rust

Actually there is, and it's called the "deref" operator. But what it means can vary by context.

https://doc.rust-lang.org/std/ops/trait.Deref.html

Reply:
Or if you were meaning pointer types, Rust has those also: https://doc.rust-lang.org/std/primitive.pointer.html

Reply:
Hmm, in that case we may want to add TyRef that is currently in comment, and use TyRef for & and TyPointer for * for Rust (anyway I will need TyRef when I port tree-sitter-cpp to the generic AST).

Reply:
Ok, I added it for now.